### PR TITLE
Add AIE_PARTITION parser

### PR DESCRIPTION
**Pull request by larry9523**
#### Problem solved by the commit
Add an parser of AIE_PARTITION section in XCLBIN

#### How problem was solved, alternative solutions (if any) and why they were rejected
Create a wrapper and a helper function to walk through AIE_PARTITION section in XCLBIN to fill required information to the wrapper and return that wrapper in the helper function

#### Risks (if any) associated the changes in the commit
No

#### What has been tested and how, request additional testing if necessary
Wrote a small test code in hw_emu to call the helper function to get the data in AIE_PARTITION with a 1x4 AIE design that can be relocated to any column on {1, 2, 3, 4}. The parsed partition metadata is correct. See below output of the test code

ncol is 1
name is xclbinutil-generated
start col 0 is 1
start col 1 is 2
start col 2 is 3
start col 3 is 4

#### Documentation impact (if any)
N/A

### CODE REVIEW

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 1]
```diff
@@ -836,6 +836,27 @@ get_softkernels(const axlf* top)
...
+  auto aiep = reinterpret_cast<const aie_partition*>(begin);
+  obj.ncol = aiep->info.column_width;
+  obj.name = std::string(begin + aiep->mpo_name);
+  auto scp = reinterpret_cast<const uint8_t*>(begin + aiep->info.mpo_auint16_start_columns);
+  for (uint32_t i = 0; i < aiep->info.start_columns_count; i++) {
+    obj.start_col_list.push_back(scp[i]);
+    scp++;
+  }
+
+  return obj;
```
Comment (stsoe) [inline src/runtime_src/core/common/xclbin_parser.cpp:857]:
```suggestion
  auto scp = reinterpret_cast<const uint8_t*>(begin + aiep->info.mpo_auint16_start_columns);
  return {obj.ncol, {scp, scp + aiep->info.start_columns_count}, begin + aiep->mpo_name};
```

**Review by stsoe — CHANGES_REQUESTED** [review]
Minor style changes.

File: src/runtime_src/core/common/xclbin_parser.h [hunk 2]
```diff
@@ -95,6 +95,18 @@ struct softkernel_object
   char *sk_buf;
 };
 
+// struct aie_partition_obj - wrapper for an AIE Partition object
+//
+// @ncol: number of columns in this partition
+// @start_col_list: Array of start column for partition relocation
+// @name: partition name
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/xclbin_parser.h:102]: ok got it.

**Review by stsoe — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 3]
```diff
@@ -836,6 +836,20 @@ get_softkernels(const axlf* top)
...
+{
+  auto pSection = ::xclbin::get_axlf_section(top, AIE_PARTITION);
+  if (!pSection)
+    return {};
+
+  auto begin = reinterpret_cast<const char*>(top) + pSection->m_sectionOffset;
+  auto aiep = reinterpret_cast<const aie_partition*>(begin);
+  auto scp = reinterpret_cast<const uint16_t*>(begin + aiep->info.mpo_auint16_start_columns);
+
+  return {aiep->info.column_width, {scp, scp + aiep->info.start_columns_count}, begin + aiep->mpo_name};
```
Comment (pranjalv-xilinx) [inline src/runtime_src/core/common/xclbin_parser.cpp:850]: trying to understand how "begin + aiep->mpo_name" is the name of partition? is mpo_name actually the offset to the actual string?

Comment (larry9523) [reply src/runtime_src/core/common/xclbin_parser.cpp:850]: Yes, anything start from mpo_ is an offset to the actual object. e.g. mpo_aunit16_start_columns

**Review by pranjalv-xilinx — APPROVED** [review]

**Pull request merged**